stdlib/fun.mli

```diff
@@ -31,3 +31,31 @@ val flip : ('a -> 'b -> 'c) -> ('b -> 'a -> 'c)
 val negate : ('a -> bool) -> ('a -> bool)
 (** [negate p] is the negation of the predicate function [p]. For any
     argument [x], [(negate p) x] is [not (p x)]. *)
+
+val protect : acquire:(unit -> 'a) -> release:('a -> unit) -> ('a -> 'b) -> 'b
+(** [protect ~acquire ~release work] invokes [acquires], then [work],
+   and then invokes [release] when [work] returns with its value or an
+   exception. In the latter case the exception is re-raised after
+   [release]. If [release] raises an exception, then the exception
+   [Release_failure] is raised instead, as documented below.
+
+    [protect] lets you manage resources reliably, under the following
+   conditions: 1) the acquisition either succeeds, or if it raises an
+   exception then it is without having acquired the resource (strong
+   exception safety); 2) the release of the resource never fails. If
+   the conditions are met, then the resource is guaranteed to have
+   been released when [protect] returns.
+
+    Note: [protect] does not yet protect against asynchronous
+   exceptions raised inside [acquire] or [release] by signal handlers,
+   such as Sys.Break. Furthermore, an asynchronous exception in
+   [release] can be promoted into a [Release_failure] exception.
```

Comment:
That would be worse—you could catch `Sys.Break` or `Out_of_memory` and not know whether the world is in a consistent state.

Reply:
I don't see how wrapping them in `Release_raised` makes things less worse.

Reply:
The discussion is about having some guarantees even in the event of `Sys.Break` and `Out_of_memory`.
I will close the review because in the new approach, `Finally_raised` has a different meaning and no longer implies a violation of resource safety (before that, I also became convinced that it should have been `exit`).